packages/datadog-plugin-couchbase/test/index.spec.js

```diff
@@ -147,124 +104,195 @@ describe('Plugin', () => {
+
+          it('should skip instrumentation for invalid arguments', (done) => {
+            const checkError = (e) => {
+              expect(e.message).to.be.oneOf([
+                // depending on version of node
+                'Cannot read property \'toString\' of undefined',
+                'Cannot read properties of undefined (reading \'toString\')',
+                'parsing failure' // sdk 4
+              ])
+              done()
+            }
+            try {
+              cluster.query(undefined).catch(checkError) // catch bad errors
+            } catch (e) {
+              // catch errors conventionally as well
+              checkError(e)
+            }
           })
         })
 
-        // after v3, buckets no longer support querying
-        // TODO: bucket viewquery?
-        withSemverGTE3(version, undefined, () => {
-          describe('queries on buckets', () => {
-            it('should handle N1QL queries', done => {
-              const query = 'SELECT 1+2'
-
-              agent
-                .use(traces => {
-                  const span = traces[0][0]
-                  expect(span).to.have.property('name', 'couchbase.query')
-                  expect(span).to.have.property('service', 'test-couchbase')
-                  expect(span).to.have.property('resource', query)
-                  expect(span).to.have.property('type', 'sql')
-                  expect(span.meta).to.have.property('span.kind', 'client')
-                  expect(span.meta).to.have.property('couchbase.bucket.name', 'datadog-test')
-                })
-                .then(done)
-                .catch(done)
-
-              const n1qlQuery = N1qlQuery.fromString(query)
-              bucket.query(n1qlQuery, (err) => {
-                if (err) done(err)
+        describe('operations still work with callbacks', () => {
+          it('should perform normal cluster query operation with callback', done => {
+            agent
+              .use(traces => {
+                const span = traces[0][0]
+                expect(span).to.have.property('name', 'couchbase.query')
+                expect(span).to.have.property('service', 'test-couchbase')
+                expect(span).to.have.property('resource', query)
+                expect(span).to.have.property('type', 'sql')
+                expect(span.meta).to.have.property('span.kind', 'client')
               })
+              .then(done)
+              .catch(done)
+
+            const query = 'SELECT 1+1'
+            cluster.query(query, (err, rows) => {
+              if (err) done(err)
+            })
+          })
+          describe('errors are handled correctly in callbacks', () => {
+            it('should catch error in callback for non-traced functions', done => {
+              const invalidIndex = '-1'
+              collection.get(invalidIndex, (err) => { if (err) done() })
             })
+
+            // due to bug in couchbase for these versions (see JSCBC-945)
```

Comment:
What happens in these versions? Does tracing simply not work or do we cause a crash?

Reply:
We don't cause a crash - the referenced bug results in the incorrect argument being read back as the callback functions in the couchbase library, so for v3.2.0-3.2.1 of couchbase, callback functions just don't work for `cluster.query`. Tracing should still be fine, since the span will finish when the promise resolves anyways, which is default behavior added in couchbase v>=3, but for testing purposes, trying to see if the callback is still called from our wrapper won't work, since the underlying `query` method doesn't work with callbacks for these versions.